Remove FILEDosToUnixPathA conversion

janvorli opened the pull request:
All file paths passed to the coreclr PAL are processed by this function to convert backslash characters to forward ones. This is causing problems when users have directories with names containing backslashes. Such problems were recently reported by people.

This change removes the `FILEDosToUnixPathA` and fixes one PAL test and one coreclr test that was passing backslash separated paths.

janvorli committed:
Remove FILEDosToUnixPathA conversion

All file paths passed to the coreclr PAL are processed by this function
to convert backslash characters to forward ones. This is causing problems
when users have directories with names containing backslashes. Such
problems were recently reported by people.

This change removes the `FILEDosToUnixPathA` and fixes one PAL test and
one coreclr test that was passing backslash separated paths.

jkotas commented:
There are a few places in the runtime that have similar problem. Could you please scan and fix them as well?

For example, https://github.com/dotnet/runtime/blob/2f9418db6b724b46b1d2358ce55c088d160a87eb/src/coreclr/utilcode/splitpath.cpp#L111

jkotas commented:
This needs to be tracked as a breaking change.

AaronRobinsonMSFT commented:
@elinor-fung Isn't there an issue that is tracking this behavior?

janvorli committed:
Remove checks for `\` on Unix and some conversions

MichalStrehovsky commented:
NativeAOT is also emulating this:

https://github.com/dotnet/runtime/blob/1076cf52f60a845cf59e316345e2e599a6fcd3ac/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.NativeAot.Unix.cs#L14-L15

janvorli committed:
Update runtime to use platform specific directory separator

janvorli commented:
@jkotas, @AaronRobinsonMSFT I have scanned the whole coreclr and nativeaot and modified all places where we were either converting the directory separators, or using both or even using backslashes on Unix too. And I've removed 
`_wsplitpath_s`.

jkotas commented on src/coreclr/utilcode/longfilepathwrappers.cpp:
```diff
@@ -697,10 +697,10 @@ HRESULT LongFile::NormalizePath(SString & path)
 BOOL LongFile::ContainsDirectorySeparator(SString & path)
 {
-    return path.Find(path.Begin(), DirectorySeparatorChar) || path.Find(path.Begin(), AltDirectorySeparatorChar);
+    return path.Find(path.Begin(), DirectorySeparatorChar);
```
Should `AltDirectorySeparatorChar` handling stay here for Windows?

janvorli replied:
This function is actually unused, I will remove it. However, the `IsDirectorySeparator` where I've also removed the alt directory handling needs to get it back, as it is used on the code path of the `LoadLibraryExWrapper` where the normalization of the path was added few years ago to handle issue with some NuGet packages containing paths with all forward slashes.

janvorli replied:
I am not sure though with all the usages of the `NormalizePath` that calls the `IsDirectorySeparator`. All the wrappers in this file that get file path as an input call the `NormalizePath`. @jkotas do you see a reason for accepting both forms of slash in methods like `GetFileAttributesWrapper`, `CreateFileWrapper` etc? I have thought the `LoadLibraryExWrapper` is an exception.

janvorli replied:
Do Windows APIs consider the slash and backslash to be directory separators? That would be surprising to me. I don't see that mentioned in the MSDN doc at https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file.

janvorli replied:
Hmm, I've tried it and they really do support that. Interesting.

jkotas commented on src/coreclr/utilcode/longfilepathwrappers.cpp:
```diff
@@ -534,8 +534,8 @@ CopyFileExWrapper(
 const WCHAR LongFile::DirectorySeparatorChar = W('\\');
-const WCHAR LongFile::AltDirectorySeparatorChar = W('/');
 #ifdef HOST_WINDOWS
+const WCHAR LongFile::AltDirectorySeparatorChar = W('/');
```
Why are we compiling any of this for Unix? Should this file be Windows-specific?

janvorli replied:
This file contains also wrappers for functions that return string in a buffer and those wrappers get SString argument. But I will ifdef out the functions that are not this case and the LongFile class.

jkotas commented on src/coreclr/vm/peimage.cpp:
```diff
@@ -593,7 +594,7 @@ void PEImage::EnumMemoryRegions(CLRDataEnumMemoryFlags flags)
 
                         //Find the filename from pdb full path
-                        char* fileName = strrchr(pCvInfo->path, '\\');
+                        char* fileName = strrchr(pCvInfo->path, DIRECTORY_SEPARATOR_CHAR_A);
```
This is searching in a path embedded in .pdb file. I am not sure whether this change is right.

AaronRobinsonMSFT replied:
If we embed on non-Windows, I would assume the paths could have `/`.

janvorli replied:
It looks like that for managed assemblies we generate PDB files in "BSJB" format. Only native Windows libraries have the RSDS format. This code path is taken only for the latter (there is an "if" a bit above). So this change needs to be reserved and we would always get files with backslash separators here.

janvorli committed:
Revert changes at few places and remove some code for Unix

janvorli committed:
Fix Bundle_Extraction_To_Relative_Path_Succeeds host test

Exclude the `foo\\bar` case for non-Windows.

janvorli commented:
The `AppHost.Bundle.Tests.BundleExtractToSpecificPath.Bundle_Extraction_To_Relative_Path_Succeeds` fails with these changes for this `[InlineData("foo\\bar", BundleOptions.BundleNativeBinaries)]`. This test case should be Windows specific after this change as it expects the foo and bar to be separate directories. We already have an exclusion check for similar case with forward slash on Windows, so I will mimic that.

janvorli commented:
@jkotas, @AaronRobinsonMSFT the CI is green and I believe I have addressed all of the feedback. I'll merge it in the morning unless you have an additional feedback.

janvorli commented:
The breaking change issue is here: https://github.com/dotnet/docs/issues/32906
